%\VignetteIndexEntry{BioQC-kidney: The kidney expression example}
%\VignetteDepends{Biobase}
%\VignettePackage{BioQC}

\documentclass[11pt]{article}

\usepackage{times}
\usepackage{hyperref}
\usepackage{geometry}
\usepackage{longtable}
\usepackage[pdftex]{graphicx}
\SweaveOpts{keep.source=TRUE,eps=FALSE,pdf=TRUE,prefix=TRUE} 

% R part
\newcommand{\R}[1]{{\textsf{#1}}}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\Metas}[1]{{\texttt{#1}}}

\begin{document}
\setkeys{Gin}{width=0.8\textwidth}
\title{Case study of BioQC: detecting tissue heterogeneity in a kidney expression dataset}
\author{Jitao David Zhang}

\maketitle

\begin{abstract}
In this vignette we demonstrate the use of \Rpackage{BioQC} with a biological dataset, where mouse kidney samples were profiled for gene expression. Results of BioQC pointed to potential tissue heterogeneity caused by pancreas, which was confirmed by qRT-PCR.
\end{abstract}

This is a supplementary documentation of \Rpackage{BioQC}, a R/Bioconductor package used to detect tissue heterogeneity from high-throughput gene expression profiling data with tissue-specific gene signatures. For its basic use please refer to the documentation and vignettes shipped along with the package. Here we demonstrate its use with a real biological data set, which is not included in the package distribution due to size limitations.

\section{Data read in}

First we load BioQC library, the tissue-specific gene signatures, and the expression data into the R session. 

<<dataReadIn>>=
library(BioQC)

gmtFile <- system.file("extdata/exp.tissuemark.affy.roche.symbols.gmt", 
                       package="BioQC")
gmt <- readGmt(gmtFile)

file <- "bioqc-nephrectomy.RData"
load(file)
print(eset)
@ 

The profile contains expression of \Sexpr{nrow(eset)} genes in \Sexpr{ncol(eset)} samples. THe expression profile was normalised with RMA normalization. The signals were \Rfunction{log2}-transformed; however, this step does not affect the result of \Rpackage{BioQC} since it uses non-parametric Wilcoxon-Mann-Whitney rank sum test.

\section{Run BioQC}
Next we run the core function of the \Rpackage{BioQC} package, \Rfunction{wmwTest}, to perform the analysis.
<<runBioQC>>=
system.time(bioqcRes <- wmwTest(eset, gmt, 
                                alternative="greater"))
@ 

The function returns \textit{one-sided} \textit{p}-values of WMW test. We next visualize this metric after transformation.
<<visBioQCprepare>>=
bioqcResFil <- filterPmat(bioqcRes, 1E-8)
bioqcAbsLogRes <- absLog10p(bioqcResFil)
@ 

<<visBioQC, include=FALSE>>=
filRes <- bioqcAbsLogRes[c("Kidney_NGS_RNASEQATLAS_0.6_3", 
                           "Pancreas_Islets_NR_0.7_3"),]
matplot(t(filRes), pch=c("K", "P"), type="b", lty=1L,
        ylab="BioQC score", xlab="Sample index")
@ 

\begin{figure}
\begin{center}
<<doVisBioQC, fig=TRUE, echo=FALSE>>=
<<visBioQC>>
@ 
\end{center}
\caption{BioQC scores (defined as $abs(log10(p))$) of the samples. K and P represent kidney and pancreas signature scores, respectively.}
\label{fig:BioQCkidney}
\end{figure}

Visual inspection (Figure~\ref{fig:BioQCkidney}) reveals that there might be sample contanminations in the samples 23-25, potentially by pancreas tissues.

\section{Biological validation}
To confirm the hypothesis generated by \Rpackage{BioQC}, we performed qRT-PCR experiments to test two pancreas-specific genes' expression in the same set of samples. Note that the two genes (amylase and elastase) are not included in the signature set provided by BioQC.

<<rtpcrRes, include=FALSE>>=
amylase <- eset$Amylase
elastase <- eset$Elastase
pancreasScore <- bioqcAbsLogRes["Pancreas_Islets_NR_0.7_3",]
par(mfrow=c(1,2), mar=c(3,3,1,1), mgp=c(2,1,0))
plot(amylase~pancreasScore, log="y", pch=21, bg="red", 
     xlab="BioQC pancreas score", ylab="Amylase")
text(pancreasScore,amylase, 1:ncol(eset), pos=1)
plot(elastase~pancreasScore, log="y", pch=21, bg="red",
     xlab="BioQC pancreas score", ylab="Elastase")
text(pancreasScore,elastase, 1:ncol(eset), pos=1)
@ 

\begin{figure}
\begin{center}
<<doRtpcrRes, fig=TRUE, echo=FALSE, width=8, height=4>>=
<<rtpcrRes>>
@ 
\end{center}
\caption{Correlation between qRT-PCR results and BioQC pancreas score}
\label{fig:rtpcr}
\end{figure}

The results are shown in Figure~\ref{fig:rtpcr}.It seems likely that sample 23-25 are contaminated by nearby pancreas tissues when the kidney was dissected.

%% add discussions: why p-value can be used?
%\section{Discussion}

\section{Acknowledgment}
I want to thank Laura Badi, Guido Steiner, Gonzalo Dur\'{a}n Pacheco, and Martin Ebeling for trying out and providing feedbacks to improve the package, Laurent Essioux for discussions, and Silvia Ines Pomposiello for expression and PCR data.

\section{Session Info}
The script runs within the following session:
<<sessionInfo, echo=FALSE, results=verbatim>>=
sessionInfo()
@ 

\end{document}
