---
title: "BioQC-kidney: The kidney expression example"
output: 
  html_document: 
    keep_md: yes
---

The BioQC-kindey example
========================
Supplementary Information for "Detect issue heterogenity in gene expression data with [*BioQC*](https://github.com/Accio/BioQC)" ([Jitao David Zhang](mailto:jitao_david.zhang@roche.com), Klas Hatje, Clemens Broger, Martin Ebeling and [Laura Badi](laura.badi@roche.com))


```{r setup, include=FALSE}
options(fig_caption=TRUE)
library(knitr)
opts_chunk$set(out.extra='style="display:block; margin: auto"', fig.align="center")
```

Introduction
------------
In this vignette, we demonstrate the use of *BioQC* with a case study where mouse kidney samples were profiled for gene expression. Results of BioQC pointed to potential tissue heterogeneity caused by pancreas contamination which was confirmed by qRT-PCR experiments. Source code and data needed to reproduce this document can be found in the github repository [BioQC-example](https://github.com/Accio/BioQC-example).

This is a supplementary documentation of *BioQC*, a R/Bioconductor package used to detect tissue heterogeneity from high-throughput gene expression profiling data with tissue-specific gene signatures. For its basic use please refer to the documentation and vignettes shipped along with the package, or to the other vignette [bioqc-simulation](bioqc-simulation.md) which applies the algorithm to simulated datsets. Here we demonstrate its use with a real biological data set, which is not included in the package distribution due to size limitations.


Importing the data
------------------
First, we load the package, the tissue-specific gene signatures, and the expression data into the R session. 

```{r lib, warning=FALSE, message=FALSE, results="hide"}
library(BioQC)

gmtFile <- system.file("extdata/exp.tissuemark.affy.roche.symbols.gmt",
                        package="BioQC")
gmt <- readGmt(gmtFile)
```

```{r load_data}
file <- "bioqc-nephrectomy.RData"
load(file)
print(eset)
```

The dataset contains expression of `r nrow(eset)` genes in `r ncol(eset)` samples. The expression profile was normalized with RMA normalization. The signals were  also `log2`-transformed; however, this step does not affect the result of *BioQC* since it is essentially a non-parametric statistical test.


Running BioQC
-------------
Next we run the core function of the *BioQC* package, `wmwTest`, to perform the analysis.

```{r run_bioqc}
system.time(bioqcRes <- wmwTest(eset, gmt,
                                 valType="p.greater"))
```

The function returns *one-sided* $p$-values of Wilcoxon-Mann-Whitney test. We next visualize this metric after transformation.

```{r visualize_bioqc_prepare}
bioqcResFil <- filterPmat(bioqcRes, 1E-8)
bioqcAbsLogRes <- absLog10p(bioqcResFil)
```

By closer examination we find that the expression of pancreas and adipose specific genes is significantly enriched in samples 23-25: 

```{r heatmap, fig.width=8, fig.height=4, dev='svg'}
library(RColorBrewer)
heatmap(bioqcAbsLogRes, Colv=NA, Rowv=TRUE,
        cexRow=0.85,
        col=rev(brewer.pal(7, "RdBu")),
        labCol=1:ncol(bioqcAbsLogRes))
```
(BioQC scores (defined as $|\log_{10}(p)|$) of the samples visualized in heatmap. Red and blue indicate high and low scores respectively.)

Visual inspection reveals that there might be contaminations in samples 23-25, potentially by pancreas and adipose tissue. 
```{r vis_bioqc, fig.width=8, fig.height=4, dev='svg', fig.cap="test"}
filRes <- bioqcAbsLogRes[c("Kidney_NGS_RNASEQATLAS_0.6_3",
                           "Pancreas_Islets_NR_0.7_3"),]
matplot(t(filRes), pch=c("K", "P"), type="b", lty=1L,
        ylab="BioQC score", xlab="Sample index")
```
(BioQC scores (defined as $|\log_{10}(p)|$) of the samples. K and P represent kidney and pancreas signature scores respectively. )

Validation with quantitative RT-PCR
===================================
To confirm the hypothesis generated by *BioQC*, we performed qRT-PCR experiments to test two pancreas-specific genes' expression in the same  set of samples. Note that the two genes (amylase and elastase) are not included in the signature set provided by BioQC.









R Session Info
----------------
```{r session_info}
sessionInfo()
```