%\VignetteIndexEntry{BioQC-kidney: The kidney expression example}
%\VignetteDepends{limma, Biobase, RColorBrewer, xtable}
%\VignettePackage{BioQC}

\documentclass[8pt]{article}

\usepackage{amsmath}
\usepackage{natbib}
\usepackage{fullpage}
\usepackage{setspace}
\usepackage{fancyvrb}
\usepackage{times}
\usepackage{hyperref}
\usepackage{geometry}
\usepackage{longtable}
\usepackage[pdftex]{graphicx}
\usepackage[usenames,dvipsnames]{color}
\DefineVerbatimEnvironment{code}{Verbatim}{fontsize=\small, xleftmargin=5mm}
\DefineVerbatimEnvironment{example}{Verbatim}{fontsize=\small}

\SweaveOpts{keep.source=TRUE,eps=FALSE,pdf=TRUE,prefix=TRUE} 

% R part
\newcommand{\R}[1]{{\textsf{#1}}}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\Metas}[1]{{\texttt{#1}}}

\hypersetup{
  colorlinks,
  citecolor=Violet,
  linkcolor=Red,
  urlcolor=Blue
}

\newcommand{\HRule}{\rule{\linewidth}{0.5mm}}
\let\endtitlepage\relax
\setstretch{1.10}

\begin{document}

\begin{titlepage}
\begin{center}

\textsc{\textsc{Supplementary information for} }\\[2em]

\begin{minipage}{0.9\textwidth}
\HRule \\[0.4cm]
\centering \large \bfseries Detect tissue heterogeneity in gene expression data with \Rpackage{BioQC} \\[0.4cm]
\HRule \\[0.4cm]
\end{minipage}

\vspace{1em}

\begin{minipage}{0.9\textwidth}
  \centering \href{mailto:jitao_david.zhang@roche.com}{Jitao David Zhang}, Klas Hatje, Clemens Broger, Martin Ebeling, and \href{laura.badi@roche.com}{Laura Badi}

\vspace{0.25em}
\centering \today
\end{minipage}

\end{center}
\end{titlepage}

\begin{abstract}
In this vignette, we demonstrate the use of \Rpackage{BioQC} with a case study where mouse kidney samples were profiled for gene expression. Results of BioQC pointed to potential tissue heterogeneity caused by pancreas contamination which was confirmed by qRT-PCR experiments. Source code and data needed to reproduce this document can be found at \url{https://github.com/Accio/BioQC-example}.
\end{abstract}

This is a supplementary documentation of \Rpackage{BioQC}, a R/Bioconductor package used to detect tissue heterogeneity from high-throughput gene expression profiling data with tissue-specific gene signatures. For its basic use please refer to the documentation and vignettes shipped along with the package, or to the other vignette \textit{bioqc-simulation.Rnw} which applies the algorithm to simulated datsets. Here we demonstrate its use with a real biological data set, which is not included in the package distribution due to size limitations.

\section{Importing data}

First we load the package, the tissue-specific gene signatures, and the expression data into the R session. 

<<dataReadIn>>=
library(BioQC)

gmtFile <- system.file("extdata/exp.tissuemark.affy.roche.symbols.gmt", 
                       package="BioQC")
gmt <- readGmt(gmtFile)

file <- "bioqc-nephrectomy.RData"
load(file)
print(eset)
@ 

The dataset contains expression of \Sexpr{nrow(eset)} genes in \Sexpr{ncol(eset)} samples. The expression profile was normalized with RMA normalization. The signals were  also \Rfunction{log2}-transformed; however, this step does not affect the result of \Rpackage{BioQC} since it is essentially a non-parametric statistical test.

\section{Running BioQC}
Next we run the core function of the \Rpackage{BioQC} package, \Rfunction{wmwTest}, to perform the analysis.
<<runBioQC>>=
system.time(bioqcRes <- wmwTest(eset, gmt, 
                                alternative="greater"))
@ 

The function returns \textit{one-sided} \textit{p}-values of Wilcoxon-Mann-Whitney test. We next visualize this metric after transformation.
<<visBioQCprepare>>=
bioqcResFil <- filterPmat(bioqcRes, 1E-8)
bioqcAbsLogRes <- absLog10p(bioqcResFil)
@ 

By closer examination (e.g. using heatmaps such as the one shown in Fig~\ref{fig:heatmap}), we found expression of pancreas and adipose specific genes is significantly enriched in samples 23-25.

<<heatmap, include=FALSE, width=8, height=3.5>>=
library(RColorBrewer)
heatmap(bioqcAbsLogRes, Colv=NA, Rowv=TRUE, 
        cexRow=0.85, margin=c(3,12),
        col=rev(brewer.pal(7, "RdBu")),
        labCol=1:ncol(bioqcAbsLogRes))
@ 

\begin{figure}
\begin{center}
<<doHeatmap, fig=TRUE, echo=FALSE>>=
<<heatmap>>
@ 
\end{center}
\caption{BioQC scores (defined as $abs(log10(p))$) of the samples visualized in heatmap. Red and blue indicate high and low scores, respectively.}
\label{fig:heatmap}
\end{figure}


<<visBioQC, include=FALSE>>=
filRes <- bioqcAbsLogRes[c("Kidney_NGS_RNASEQATLAS_0.6_3", 
                           "Pancreas_Islets_NR_0.7_3"),]
matplot(t(filRes), pch=c("K", "P"), type="b", lty=1L,
        ylab="BioQC score", xlab="Sample index")
@ 

\begin{figure}
\begin{center}
<<doVisBioQC, fig=TRUE, echo=FALSE>>=
<<visBioQC>>
@ 
\end{center}
\caption{BioQC scores (defined as $abs(log10(p))$) of the samples. K and P represent kidney and pancreas signature scores, respectively.}
\label{fig:BioQCkidney}
\end{figure}

Visual inspection (Figure~\ref{fig:BioQCkidney}) reveals that there might be contaminations in samples 23-25, potentially by pancreas and adipose tissues.

\section{Validation with quantitative RT-PCR}
To confirm the hypothesis generated by \Rpackage{BioQC}, we performed qRT-PCR experiments to test two pancreas-specific genes' expression in the same set of samples. Note that the two genes (amylase and elastase) are not included in the signature set provided by BioQC.

<<rtpcrRes, include=FALSE>>=
amylase <- eset$Amylase
elastase <- eset$Elastase
pancreasScore <- bioqcAbsLogRes["Pancreas_Islets_NR_0.7_3",]
par(mfrow=c(1,2), mar=c(3,3,1,1), mgp=c(2,1,0))
plot(amylase~pancreasScore, log="y", pch=21, bg="red", 
     xlab="BioQC pancreas score", ylab="Amylase")
text(pancreasScore,amylase, 1:ncol(eset), pos=1)
plot(elastase~pancreasScore, log="y", pch=21, bg="red",
     xlab="BioQC pancreas score", ylab="Elastase")
text(pancreasScore,elastase, 1:ncol(eset), pos=1)
@ 

\begin{figure}
\begin{center}
<<doRtpcrRes, fig=TRUE, echo=FALSE, width=8, height=4>>=
<<rtpcrRes>>
@ 
\end{center}
\caption{Correlation between qRT-PCR results and BioQC pancreas score}
\label{fig:rtpcr}
\end{figure}

The results are shown in Figure~\ref{fig:rtpcr}. It seems likely that sample 23-25 are contaminated by nearby pancreas tissues when the kidney was dissected. Potential contamination by adipose tissues remains to be tested.

\section{Impact of sample removal on differential gene expression analysis}
In this study, four mice of the \textit{FVB/NJ} strain received nephrectomy operation and treatment of Losartan, an angiotensin II receptor antagonist drug, and four mice reveiced an sham operation and Losartan. Among the Nephrectomy+Losartan group, one sample (index 24) is possibly contaminated by pancreas. Suppose now we are interested in the differential gene expression between the conditions. We now run the analysis twice, once with and once without the contaminated sample, to study the impact of removing heterogenous samples detected by \Rpackage{BioQC}.

<<deg, include=FALSE>>=
library(limma)
isNeph <- with(pData(eset), Strain=="FVB/NJ" & TREATMENTNAME %in% c("Nephrectomy-Losartan", "Sham-Losartan"))
isContam <- with(pData(eset), INDIVIDUALNAME %in% c("BN7", "FNL8", "FN6"))
esetNephContam <- eset[,isNeph]
esetNephExclContam <- eset[, isNeph & !isContam]
getDEG <- function(eset) {
  group <- factor(eset$TREATMENTNAME, levels=c("Sham-Losartan","Nephrectomy-Losartan"))
  design <- model.matrix(~group)
  colnames(design) <- c("ShamLo", "NephLo")
  contrast <- makeContrasts(contrasts="NephLo", levels=design)
  exprs(eset) <- normalizeBetweenArrays(log2(exprs(eset)))
  fit <- lmFit(eset, design=design)
  fit <- contrasts.fit(fit, contrast)
  fit <- eBayes(fit)
  tt <- topTable(fit, n=nrow(eset))
  return(tt)
}

esetNephContam.topTable <- getDEG(esetNephContam)
esetNephExclContam.topTable <- getDEG(esetNephExclContam)
esetFeats <- featureNames(eset)
esetNephTbl <- data.frame(feature=esetFeats,
                          GeneSymbol=esetNephContam.topTable[esetFeats,]$GeneSymbol,
                          OrigGeneSymbol=esetNephContam.topTable[esetFeats,]$OrigGeneSymbol,
                          Contam.logFC=esetNephContam.topTable[esetFeats,]$logFC,
                          ExclContam.logFC=esetNephExclContam.topTable[esetFeats,]$logFC)
par(mfrow=c(1,1), mar=c(3,3,1,1)+0.5, mgp=c(2,1,0))
with(esetNephTbl, smoothScatter(Contam.logFC~ExclContam.logFC, 
                                xlab="Excluding one contaminating sample [logFC]",
                                ylab="Including one contaminating sample [logFC]"))
abline(0,1)
isDiff <- with(esetNephTbl, abs(Contam.logFC-ExclContam.logFC)>=2)
with(esetNephTbl, points(Contam.logFC[isDiff]~ExclContam.logFC[isDiff], pch=16, col="red"))
diffTable <- esetNephTbl[isDiff,]
diffGenes <- unique(diffTable[,"GeneSymbol"])
diffGenesPancreas <- diffGenes %in% gmt[["Pancreas_Islets_NR_0.7_3"]]$genes
@ 

We found that \Sexpr{nrow(diffTable)} probesets mapped to \Sexpr{length(diffGenes)} are associated with very strong expression changes if the contaminated sample is not excluded (Table~\ref{tbl:diffTbl}). Not surprisingly \Sexpr{sum(diffGenesPancreas)} genes among them are highly enriched in pancreas but not kidney and belong to the pancreas signature used by \Rpackage {BioQC}.

\begin{figure}
\begin{center}
<<doDeg, fig=TRUE, echo=FALSE, width=8, height=4>>=
<<deg>>
@ 
\end{center}
\caption{Log2 fold change (logFC) values reported by \textit{limma} with one contaminated sample included (y-axis) or excluded (x-axis). Genes strongly affected by the contamination are indicated by red dots.}
\label{fig:rtpcr}
\end{figure}

\begin{table}
<<diffTbl, results=tex, echo=FALSE>>=
library(xtable)
diffXtable <- xtable(diffTable)
print(diffXtable, floating=FALSE, include.rownames=FALSE)
@ 
\label{tbl:diffTbl}
\caption{Genes that are identified as strongly changed ony if the contaminated sample is included.}
\end{table}

In summary, we observe that tissue heterogeneity can impact down-stream analysis results and negatively affect reproducibility of gene expression data if it remains overlooked. It underlines again the value of applying \Rpackage{BioQC} as a first-line quality control tool.

\section{Session Info}
The script runs within the following session:
<<sessionInfo, echo=FALSE, results=verbatim>>=
sessionInfo()
@ 

\end{document}
