---
title: Deriving tissue signatures from the GTEx dataset. 
output:
  html_document: default 
---

The *GTEx-project*[^1] is a data resource providing high quality gene expression data from multiple reference human tissues. In this document, we identify differentially expressed genes for each tissue using [*voom*](http://bioconductor.org/packages/release/bioc/html/limma.html) and 
[*limma*](http://bioconductor.org/packages/release/bioc/html/limma.html) to build tissue-specific signatures. 

These signatures can be used with [*BioQC*](https://accio.github.io/BioQC/) to detect tissue heterogeneity in gene expression data. 

[^1]: Lonsdale, J., Thomas, J., Salvatore, M., Phillips, R., Lo, E., Shad, S., … Moore, H. F. (2013). The Genotype-Tissue Expression (GTEx) project. Nature Genetics, 45(6), 580–585. http://doi.org/10.1038/ng.2653

```{r load_libs, include=FALSE}
library(ribiosUtils)
libordie(readr)
libordie(ribiosUtils)
libordie(ribiosIO)
libordie(biomaRt)
libordie(edgeR)
libordie(data.table)
libordie(stringr)
```

### Load the data
We read the raw counts from the `.gct` file and preprocessed tissue annotations from `GTEx-UDISDataSetID5681-sampleAnnotation.txt`. 
```{r read_data}
gct = read_gct_matrix("GTEX_v6/GTEx_Analysis_v6_RNA-seq_RNA-SeQCv1.1.8_gene_reads.gct")
annotation = read_tsv("data/GTEx-UDISDataSetID5681-sampleAnnotation.txt")
annotation.tissue = read_tsv("GTEX_v6/v6_UDIS.txt")
ids = colnames(gct)
annotation.matched = matchColumn(ids, annotation, "Experiment name")
annotation.tissue.matched = matchColumn(ids, annotation.tissue, "SAMPID")

colData = cbind(annotation.matched, annotation.tissue.matched)
rownames(colData) = colData[["Experiment name"]]

#filter rows with NA
# RIN = RNA quality index
# SMTS = tissue
validInds = which(!apply(is.na(colData[,c("RIN", "Gender", "SMTS")]), 1, sum) >= 1)
gct = gct[,validInds]
colData = colData[validInds,]
```

### perform differential analysis
We run a differntial expression analysis on the GTEx dataset with
```{r edgeR}
# RIN = RNA quality index
# SMTS = tissue
dgList = DGEList(counts=gct, genes=rownames(gct))
designMat = model.matrix(~colData$SMTS + colData$RIN + colData$Gender)

CPM = cpm(dgList)
cpmCnt = apply(CPM > 1, 1, sum)
cpmInds = cpmCnt >= 10
dgList.f = dgList[cpmInds,]

dgList.f = calcNormFactors(dgList.f)


v = voom(dgList.f, designMat, plot=FALSE)

fit = lmFit(v, designMat)
save(fit, file="voom-limma-fit.Rdata")
fit.e = treat(fit, lfc=log2(100))

rslt = decideTests(fit.e, p.value=(0.01 / nrow(dgList.f)))


```

```{r annotation}
ensembl = useMart("ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl", host="www.ensembl.org")  
ensemble2hgnc = data.table(getBM(mart=ensembl, attributes=(c("ensembl_gene_id", "hgnc_symbol"))))
ensemble2hgnc = ensemble2hgnc[hgnc_symbol!='',]

salitizeEnid = function(x) { 
  return(str_split(x, "\\.")[[1]][1])
}

rslt.tab = data.table(rslt, keep.rownames=TRUE)
names = make.names(colnames(rslt.tab))
setnames(rslt.tab, names)
rn.2 = lapply(rslt.tab[,rn], salitizeEnid)
rslt.tab = rslt.tab[,rn2:=rn.2]

rslt.matched = data.table(matchColumn(ensemble2hgnc$ensembl_gene_id, rslt.tab, 'rn2'))
rslt.matched = rslt.matched[,hgnc:=ensemble2hgnc[,hgnc_symbol]]

signatures = list()
sigCols =  names[grepl("^colData", names)]
sigCols = sigCols[1:(length(sigCols)-2)] # remove RIN and Gender
for(tCol in sigCols) {
  signatures = append(signatures, list(rslt.matched[get(tCol)==1,hgnc]))
}
names(signatures) = sigCols

save(signatures, file="gtex_signatures.Rdata")
```


